test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageBuilderTest.cs

```diff
@@ -186,7 +240,7 @@ public void CreatePackageWithNuspecIncludeExclude()
       <group targetFramework="".NETFramework4.5"">
         <dependency id=""packageB"" version=""1.0.0"" exclude=""z"" />
       </group>
-      <group targetFramework="".NETFramework4.6"">
+      <group targetFramework=""net5.0-windows"">
```

Comment:
nuspecs should definately never allow a net5.0 TFM with a TPI, but no TPV.

Reply:
This is actually totally valid! There's going to be a default version provided by the sdk for each major .net version

Reply:
My understanding is that it's valid for a project file (`csproj`). But in the future when we restore a project targeting `net6.0-windows10.0.12345`, and the package has binaries for `net5.0-windows`, we won't know what the default platform version was for windows for `net5.0-windows` at the time that the package author packed the project.

So, the `csproj` uses `net5.0-windows`, but restore gets the default version from the SDK, and writes that in the assets file. I'm not sure if pack reads the project TFMs from the assets file, or if it also needs to get the default platform version from the SDK, but it's resolved at pack time and burned into the nuspec and the package.

Otherwise every year, the SDK will need to pass an array of every previous .net's version default platform versions. Imagine `net20` trying to figure out what the `net5.0-windows` default platform version was.

Reply:
I think @zivkan brings out a good point. 
Pack needs to hard code the version here.

Reply:
....or maybe I misunderstood :( I guess this is THE place where we shouldn't allow it?

Reply:
my understanding is that the csproj TargetFramework(s) is the only place where a TFM with a TPI but no TPV is allowed, and that's because the SDK will provide the default value.  Therefore anything else that cannot get the default TPV from the SDK cannot treat the TPV as optional. At least, that's my understanding.

I believe the SDK is responsible for parsing TargetFramework into TFI, TFV, TPI, and either parsing the TPV or generating the default value. This will only work if NuGet stops parsing the TargetFramework directly, and starts using the other properties, which is large the "tfm aliasing" work that we also have planned.

So, I think this PR needs to create APIs to enable TFM with TPI and TPV, but I guess we can't support optional TPV until we complete that initial aliasing work. At least my opinion.  Maybe let's schedule a call tomorrow to clear things up more quickly.

Reply:
@nkolev92 @zivkan @zkat The plan is that the folder name would not include the TPV, but that information would be included in the manifest.  See https://github.com/dotnet/designs/blob/master/accepted/2020/net5/net5.md#nuget-pack-behavior, though note that the specific format for this information is mostly up to NuGet.

We don't want to block this PR on completing the updates to the manifest, pack, etc. though.  We see that as the next step after this PR merges.

Reply:
The folder name and the dependencies elements are looked at independently. 

I think the folder name in the package itself will need to contain the min version. It'll be inferred as 0.0 if not provided.

Reply:
We have a meeting invite for later today, so I hope we can discuss it more quickly than these comments going back and forth, but @dsplaisted's comment in another issue: https://github.com/nuget/home/issues/9444#issuecomment-613761628  appears to be evidence that once NuGet starts reading the 4/5 MSBuild properties, rather than parsing TargetFramework directly, by the time we get to C# code, TPV will always be defined. My understanding is still that the c# class `NuGet.Frameworks.NuGetFramework` can treat TPV as mandatory when TPI is not the empty string/null.

Reply:
> My understanding is still that the c# class `NuGet.Frameworks.NuGetFramework` can treat TPV as mandatory when TPI is not the empty string/null.

@zivkan I don't think that will work.  We're going to be using NuGetFramework to parse out those MSBuild properties from the `TargetFramework`.  For example (from [this PR](https://github.com/dotnet/sdk/pull/12202)):

```xml
    <TargetPlatformIdentifier>$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)'))</TargetPlatformIdentifier>
    <TargetPlatformVersion>$([MSBuild]::GetTargetPlatformVersion('$(TargetFramework)', 3))</TargetPlatformVersion>
```

Those intrinsic MSBuild functions [will be implemented](https://github.com/microsoft/msbuild/pull/5429) by calling the NuGetFramework APIs, so they need to work when you have the platform name but not the version.

You might be able to require that the platform version be set if the platform name is set when doing compatibility checks for a NuGetFramework.

Reply:
I see. I was reading the SDK's master branch targets file, which does string manipulation, so thought the SDK wouldn't be using our API. This indeed makes things more complex, but for a very good reason.

Reply:
@zkat @nkolev92 is adding a new `ParseProjectTargetFramework(string)` that allows optional TPV, but all the other `Parse*` methods keeping it mandatory a horrible idea?

Maybe I should give up on this idea and accept that the rest of our code is going to have higher risk of bugs if we don't check if TPV is defined correctly.

Reply:
Apparently `NuGetFramework.Parse("netcoreapp")` works and returns an object with TargetFrameworkfVersion=0.0. So maybe this isn't a good suggestion.  I feel like I needlessly delayed this PR now.

I think it would be good for the nuspec and nupkg's `lib/tfm` directory to contain the version 0.0 TPV. It's a signal that something went wrong. I don't know if `GetShortFolderName` should always return TPV, or if we should limit it to pack. My feeling is always, but I don't know if `netcoreapp0.0` or `netcoreapp` is returned by `GetShortFoldName` in the example above.